specs/core/0_beacon-chain.md

```diff
@@ -1628,74 +1582,55 @@ def process_crosslinks(state: BeaconState) -> None:
-    return get_base_reward(state, index) + extra_penalty
 ```
 
 ```python
-def get_justification_and_finalization_deltas(state: BeaconState) -> Tuple[List[Gwei], List[Gwei]]:
-    current_epoch = get_current_epoch(state)
-    epochs_since_finality = current_epoch + 1 - state.finalized_epoch
-    rewards = [0 for index in range(len(state.validator_registry))]
-    penalties = [0 for index in range(len(state.validator_registry))]
-    # Some helper variables
-    boundary_attestations = get_previous_epoch_boundary_attestations(state)
-    boundary_attesting_balance = get_attesting_balance(state, boundary_attestations)
-    total_balance = get_previous_total_balance(state)
-    total_attesting_balance = get_attesting_balance(state, state.previous_epoch_attestations)
-    matching_head_attestations = get_previous_epoch_matching_head_attestations(state)
-    matching_head_balance = get_attesting_balance(state, matching_head_attestations)
-    eligible_validators = [
+def get_attestation_deltas(state: BeaconState) -> Tuple[List[Gwei], List[Gwei]]:
+    previous_epoch = get_previous_epoch(state)
+    total_balance = get_total_active_balance(state)
+    eligible_validator_indices = [
         index for index, validator in enumerate(state.validator_registry)
-        if (
-            is_active_validator(validator, current_epoch) or
-            (validator.slashed and current_epoch < validator.withdrawable_epoch)
-        )
+        if is_active_validator(validator, previous_epoch) or (validator.slashed and previous_epoch < validator.withdrawable_epoch)
     ]
-    # Process rewards or penalties for all validators
-    for index in eligible_validators:
+    rewards = [0 for index in range(len(state.validator_registry))]
+    penalties = [0 for index in range(len(state.validator_registry))]
+    for index in eligible_validator_indices:
         base_reward = get_base_reward(state, index)
-        # Expected FFG source
-        if index in get_unslashed_attesting_indices(state, state.previous_epoch_attestations):
-            rewards[index] += base_reward * total_attesting_balance // total_balance
-            # Inclusion speed bonus
-            earliest_attestation = get_earliest_attestation(state, state.previous_epoch_attestations, index)
+
+        # Micro-incentives for attestations matching FFG source, FFG target, and head
+        for attestations in (
+            get_matching_source_attestations(state, previous_epoch),
+            get_matching_target_attestations(state, previous_epoch),
+            get_matching_source_attestations(state, previous_epoch),
+        ):
+            if index in get_unslashed_attesting_indices(state, attestations):
+                rewards[index] += base_reward * get_attesting_balance(state, attestations) // total_balance
+            else:
+                penalties[index] += base_reward
+
+        if index in get_unslashed_attesting_indices(state, get_matching_source_attestations(state, previous_epoch)):
```

Comment:
This section doesn't work. `get_beacon_proposer_index` can only used for slots in the current epoch so this section needs to be revamped.

You can instead select all attestations with an `inclusion_slot` from the current epoch and then see if it is the first to be included for the validator index from either the previous_epoch_attestations or current (depending on slot of the attestation), and reward accordingly.

Reply:
I opted to add `proposer_index` to `PendingAttestation` similarly to `inclusion_slot`.